### ISSUE TITLE
Smartsig signer within the dApp

### ISSUE THREAD

**Issue by scholtz** [algorand/go-algorand#4920]
## Problem

AVM is capable of handling inner transactions. The only signer which is currently available is the current application signer which can sign inner app calls, payment tx, asset transaction but only from the application account or the account rekeyed to the application account.

Please add support for the smart sig signer.

We are building dApp which depends on the specifications of the other projects. The goal is to deploy smart contract application which takes some deposit and combines multiple ecosystem projects application and performs actions as it would be done by the SDK. The issue with some applications is that it requires smart sigs as part of the grouped transactions. For example AlgoFi V2 protocol requires signature of this account in order to allow landing: https://algoexplorer.io/address/ITG4MVMAQXLVEWLUH5KGRHSPVQUVHDTOCHN2WD2NZXZVZWWXHSJPAX4UEU

There are dozens of other examples, for example AlgoDex, ARC0017, ... If developers could combine freely different algorand protocols, it would be great boost for the ecosystem and DeFi growth.

## Solution

Please create new opcode such as ```itxn_sig``` which would allow developers to specify the signature program.

## Dependencies

No

## Urgency

I expect this feature to be available within 6 to 12 months.

**Comment by jannotti** [comment]
Some questions that need to be answered.  I'm hoping that people who care for such a feature will weigh in.

- Is it ok that these logicsigs will have a _significantly_ lowerer opcode budget than top-level logicsigs? (Because top-level logicsigs can be evaluated concurrently, and on the static top-level transaction, they are offered high budget: 20,000. That will not be possible when evaluating a logicsig on a dynamically generated inner transaction.
- Are _delegated_ logicsigs required, or just the simpler type (code only)? For the same reason as above, performing the signature check dynamically would require a lot more effort. (We have an internal design idea, and it's a big change - the delegated logicsigs would have to be presented in the top-level txn, even though they will get used later.)
- Is it reasonable and expected that the logicsig budget will come out of the calling app's opcode budget?
- Should we restrict such logicsigs the way logicsigs are normally restricted?  Those restrictions (like reading state) are important to allow the high budgets that logicsigs get. But in this case, we can't give them a high budget anyway. So we could run them in "app mode" and give them access to state.

**Comment by pbennett** [comment]
1, 3) There are definitely cases with legacy contracts where this would be extremely useful, and if higher opcode budgets are required and dummy transactions to bump it, so be it.  Coming out of the apps opcode budget would be the most logical to me.  An itxn signed w/ lsig would basically just have a high opcode cost.  The code constructing the lsig is a separate matter entirely.  😀 

I think the obvious longer term play is for dApps to migrate to solutions not requiring lsigs (box storage helps here) but that will require some dapps to migrate everything if they have locked contracts.  Since lsigs are still a fundamental part of the protocol though, it makes sense for contracts to be able to issue transactions using them.

**Comment by algoanne** [comment]
One idea that might make this interesting long term (beyond for supporting old designs) - if we had inner logic sigs an app could sign for a user. For example, to opt a user into an ASA.

**Comment by joe-p** [comment]
You can already sort of do this with delegated logic sigs and apps. Have a delegated lsig that allows opt ins, but only if it's called atomically with a specific app

**Comment by emg110** [comment]
Yes, @joe-p ! I agree, and open an issue on this for PeraWallet and get objected about how dangerous this feature is and given the risk it will not likely to be a feature on any wallet despite it is on Node and SDK!! Been there, done that , long ago!
Many great things are already missed from dApps because wallet providers do not support stateless smart contract signing, IMHO!
I am referring to the power of a combination of stateless and stateful smart contracts!
BTW, wouldn't it be an inconsistency in product lifeline if the parent feature never gets into wallet providers features list but after a while on-chain opcode child version of same feature comes up! 
Please correct me if there are differences that mandate such different approaches to a feature that I cannot comprehend! Million thanks!
